topicexplorer/prep.py

```diff
@@ -299,79 +300,68 @@ def get_high_filter(c, words=None, items=None, counts=None):
-                print((u'\u2588' * int(percentage * 36)).ljust(36), end=' ')
-                print("  {0:0.0f} words".format((filtered_counts >= bin).sum()).rjust(14), end=' ')
-                print(">= {0:>5.0f}x".format(bin).ljust(8))
-
-            last_row = (filtered_counts >= bin).sum()
-
-        print(' ' * 17, "{} total occurrences".format(counts.sum()).ljust(36), end=' ')
-        print('{} words total'.format(get_mask(c, words).sum()).rjust(20))
-        print('')
+    chart = ""
+
+    bin_counts, bins = np.histogram(counts, bins=bins)
+    chart += "{0:>8s} {1:>8s} {2:<36s} {3:>14s} {4:>8s}".format("Rate", 'Top', '% of corpus', "# words", "Rate") + "\n"
+    last_row = 0
+    for bin, count in zip(bins[-2::-1], np.cumsum(bin_counts[::-1])):
+        filtered_counts = counts[get_mask(c, words)]
+        if (filtered_counts >= bin).sum() > last_row:
+            percentage = 1. - (old_div(counts[counts < bin].sum(), float(c.original_length)))
+            chart += "{0:>5.0f}x".format(bin).rjust(8)
+            chart += '{0:2.1f}% '.format(percentage * 100).rjust(10)
+            chart += (u'\u2588' * int(percentage * 36)).ljust(36)
+            chart += "{0:0.0f} words".format((filtered_counts >= bin).sum()).rjust(15)
+            chart += " >={0:>5.0f}x".format(bin).ljust(8) + "\n"
+
+        last_row = (filtered_counts >= bin).sum()
+
+    chart += (' ' * 18) + "{} total occurrences".format(counts.sum()).ljust(34)
+    chart += '{} words total'.format(get_mask(c, words).sum()).rjust(20)
+    return chart
+
+def get_high_filter_stops(c, words=None, items=None, counts=None, num=None):
+    import numpy as np
+    input_filter = num
+    valid = True
+    f = open("test.txt", "w+")
+    f.write("c" + "\n")
+    f.write(str(c) + "\n")
+    f.write("c words" + "\n")
+    f.write(str(c.words) + "\n")
+    f.write("words" + "\n")
+    f.write(str(words) + "\n")
+    f.write("items" + "\n")
+    f.write(str(items) + "\n")
+    f.write("counts" + "\n")
+    f.write(str(counts) + "\n")
+    f.write("num" + "\n")
+    f.write(str(num) + "\n")
+    f.write("lengths" + "\n")
+    f.write(str(len(c.words)) + " " + str(len(items)) + " " + str(len(counts)))
+    try:
+        candidates = get_candidate_words(c, input_filter, words=words, items=items, counts=counts)
+        places = np.in1d(c.words, candidates)
+        places = dict(zip(candidates, np.where(places)[0]))
+        candidates = sorted(candidates, key=lambda x: counts[places[x]], reverse=True)
+        filtered_counts = counts[get_mask(c, words)]
+
+        filtered = ""
+        filtered += "Filter will remove " + str(filtered_counts[filtered_counts >= input_filter].sum())
+        filtered += " occurrences " + "of these " + str(len(filtered_counts[filtered_counts >= input_filter])) + " words: "
+        filtered += u' '.join(candidates)
+
+        if len(candidates) == len(c.words):
+            valid = False
```

Comment:
raise a ValueError here, rather than returning an invalid state. Our style should favor exceptions that are explicitly handled, rather than letting it silently fail and introduce inconsistency later.